Sources/NIOHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,267 @@
+// See CONTRIBUTORS.txt for the list of SwiftNIOHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOTLS
+import NIOHTTP1
+import NIOHTTP2
+import NIOConcurrencyHelpers
+import Foundation
+
+class ConnectionPool {
+    
+    init(group: EventLoopGroup, configuration: HTTPClient.Configuration) {
+        self.eventLoopGroup = group
+        self.configuration = configuration
+    }
+    
+    var eventLoopGroup: EventLoopGroup
+    var configuration: HTTPClient.Configuration
+    private var connectionProviders: [Key:ConnectionProvider] = [:]
+    private var lock = Lock()
+    
+    private subscript(key: Key) -> ConnectionProvider? {
+        get {
+            return lock.withLock {
+                connectionProviders[key]
+            }
+        }
+        set {
+            lock.withLock {
+                connectionProviders[key] = newValue
+            }
+            
+        }
+    }
+    
+    func getConnection(for request : HTTPClient.Request) -> EventLoopFuture<Connection> {
+        guard let key = makeKey(for: request.url) else {
+            fatalError("Request URL is incorrect")
+        }
+        
+        let timeout = configuration.timeout
+        
+        if let provider = self[key] {
+            return lock.withLock {
+                provider.getConnection(for: request)
+            }
+        } else {
+            var protocolChoosenPromise = eventLoopGroup.next().makePromise(of: Void.self)
+            let bootstrap = ClientBootstrap.makeHTPPClientBootstrapBase(group: self.eventLoopGroup, request: request, configuration: configuration) { channel in
+                
+                let http1PipelineConfigurator: (ChannelPipeline) -> EventLoopFuture<Void> = { pipeline in
+                    defer {
+                        protocolChoosenPromise.succeed(())
+                    }
+                    let http1Provider = HTTP1ConnectionProvider(group: self.eventLoopGroup, configuration: self.configuration, key: key)
+                    http1Provider.release(connection: Connection(key: key, channel: channel))
+                    self[key] = .http1(http1Provider)
+                    let encoder = HTTPRequestEncoder()
+                    let decoder = ByteToMessageHandler(HTTPResponseDecoder(leftOverBytesStrategy: .forwardBytes))
+                    let future = pipeline.addHandlers([encoder, decoder])
+                    return future
+                }
+                
+                let h2PipelineConfigurator: (ChannelPipeline) -> EventLoopFuture<Void> = { pipeline in
+                    let http2Handler = NIOHTTP2Handler(mode: .client, initialSettings: nioDefaultSettings)
+                    let multiplexer = HTTP2StreamMultiplexer(mode: .client, channel: channel)
+                    return pipeline.addHandlers([http2Handler, multiplexer]).map { _ in
```

Comment:
When I saw the configurators were taking a `ChannelPipeline` and not a `Channel`, I wondered if it was right to directly modify the (captured) `Channel `, so I preferred to just use the `ChannelPipeline` directly before taking time to see if it was right or not.
Why is `addHTTPClientHandlers` on `ChannelPipeline`, but not `configureHTTP2Pipeline`?

Reply:
Fixed!

Reply:
The reason `configureHTTP2Pipeline` is on the `Channel` is because configuring the multiplexer requires access to the parent `Channel`, which you cannot get from the `ChannelPipeline`.